serviceadapter-core/src/main/proto/ServiceAdapterService.proto

```diff
@@ -0,0 +1,55 @@
+syntax = "proto3";
+
+// This is the service that is used to exchange data between
+// BundleServer and the ServiceAdapters
+
+option java_multiple_files = true;
+option java_package = "net.discdd.grpc";
+
+service ServiceAdapterService {
+  // exchangeADUs sends ADUs received by the BundleServer to the ServiceAdapter and
+  // the same time receives ADUs to send back to clients as the
+  // return value.
+  rpc exchangeADUs(ExchangeADUsRequest) returns (ExchangeADUsResponse) {}
+  // check if the ServiceAdapter has any pending data to send to the BundleServer
+  // Note: The ServiceAdapter should may say that it has data for a client, but the BundleServer
+  //       may decide not to request anything. The reason for not requesting will probably be that
+  //       the client send window is full.
+  rpc pendingDataCheck(PendingDataCheckRequest) returns (PendingDataCheckResponse) {}
+}
+
+message PendingDataCheckRequest {
+    // Nothing here for now...
+}
+
+message PendingDataCheckResponse {
+    // The clientIds of the clients that have pending data
+    repeated string clientId = 1;
+}
+
+message ExchangeADUsRequest {
+  // The clientId of the client that sent the ADUs
+  string clientId = 1;
+  // ADUs being sent
+  repeated AppDataUnit adus = 2;
+  // This is the id of the ADU that was received by the BundleServer from the service adapter
+  int64 lastADUIdReceived = 3;
+}
+
+message ExchangeADUsResponse {
+  // ADUs returned from the ServiceAdapter for the clientId in the request
+  repeated AppDataUnit adus = 1;
+  //    This is the id of the ADU that was received by the adapter from the BundleServer
+  int64 lastADUIdReceived = 2;
```

Comment:
Why do we need lastADUIdReceived on both request and response? Server should be able to figure out what is the lastADUIdReceived by the Adapter. I thought the App Adapter will return all the ADUIds > request.getLastADUIdReceived() for a request and server will iterate over the ADUs and maintain the lastADUIdReceived for its next response?

Reply:
In general we don't need the lastADUIdReceived, but there could be failures, such that things get partially processed. lastADUReceived allows us to make sure to resend data that hasn't been fully processed. For example, the adapter might send 1000 ADUs to the BundleServer, but the BundleServer may only process the first 500, so later when the server talks to the adapter, the adapter will send the last 500.